lang/java/reef-utils/src/test/java/org/apache/reef/util/MultiAsyncToSyncTest.java

```diff
@@ -0,0 +1,240 @@
+ */
+final class AsynchronousIncrementer implements Callable<Integer> {
+  private static final Logger LOG = Logger.getLogger(MultiAsyncToSync.class.getName());
+  private final int sleepTimeMillis;
+  private final int input;
+  private final long identifier;
+  private final MultiAsyncToSync blocker;
+
+  /**
+   * Instantiate an incrementer with specific job parameters.
+   * @param input The input parmeter for the work.
+   * @param identifier The identifier of the caller to wake on completion.
+   * @param sleepTimeMillis How long to work.
+   * @param blocker The MultiAsyncToSync object which is holding the blocked client.
+   */
+  AsynchronousIncrementer(final Integer input, final long identifier,
+                                 final int sleepTimeMillis, final MultiAsyncToSync blocker) {
+    this.sleepTimeMillis = sleepTimeMillis;
+    this.input = input;
+    this.identifier = identifier;
+    this.blocker = blocker;
+  }
+
+  /**
+   * Sleep and then ncrement the input value by one.
+   * @return The input value of the operation incremented by one.
+   * @throws Exception
+   */
+  public Integer call() throws Exception {
+    LOG.log(Level.INFO, "Sleeping...");
+    Thread.sleep(sleepTimeMillis);
+    LOG.log(Level.INFO, "Releasing caller...");
+    blocker.release(identifier);
+    return input + 1;
+  }
+}
+
+/**
+ * Use the MultiAyncToSync class to implement a synchronous API
+ * that uses asynchronous processing internally.
+ */
+final class SynchronousApi {
+  private static final Logger LOG = Logger.getLogger(SynchronousApi.class.getName());
+  private final int incrementerSleepTimeMillis;
+  private final MultiAsyncToSync blocker;
+  private final ExecutorService executor;
+  private final ConcurrentLinkedQueue<FutureTask<Integer>> taskQueue = new ConcurrentLinkedQueue<>();
+  private AtomicLong idCounter = new AtomicLong(0);
+
+  /**
+   * Parameterize the object as to length of processing time and call timeout.
+   * @param incrementerSleepTimeSeconds Length of time the incrementer sleeps before
+   *                                    performing the increment and returning.
+   * @param timeoutPeriodSeconds The length of time before the call will timeout.
+   */
+  SynchronousApi(final int incrementerSleepTimeSeconds, final long timeoutPeriodSeconds) {
+    this.incrementerSleepTimeMillis = 1000 * incrementerSleepTimeSeconds;
+    this.blocker = new MultiAsyncToSync(timeoutPeriodSeconds, SECONDS);
+    this.executor = Executors.newFixedThreadPool(2);
+  }
+
+  /**
+   * The one API call in our interface.
+   * @param input
+   * @return The input parameter incremented by one or zero for a timeout.
+   */
+  public Integer apiCall(final Integer input) {
+    // Create a future to run the asynchronous processing.
+    final long identifier = idCounter.getAndIncrement();
+    final FutureTask<Integer> task =
+        new FutureTask(new AsynchronousIncrementer(input, identifier, incrementerSleepTimeMillis, blocker));
```

Comment:
As I mentioned above because the generic parameter of Java's FutureTask class has to be a subclass of object.